tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,235 @@
+
+    return {false, {ret}};
+}
+
+fizzy::execution_result wasi_environ_sizes_get(
+    fizzy::Instance& instance, std::vector<uint64_t> args, int)
+{
+    uint32_t environc = static_cast<uint32_t>(args[0]);
+    // TODO: implement properly (only returns 0 now)
+    uvwasi_serdes_write_uint32_t(instance.memory->data(), environc, 0);
+    return {false, {UVWASI_ESUCCESS}};
+}
+
+// TODO: support both wasi_snapshot_preview1 and wasi_unstable
+const std::vector<fizzy::ImportedFunction> wasi_functions = {
+    {"wasi_snapshot_preview1", "proc_exit", {fizzy::ValType::i32}, std::nullopt, wasi_proc_exit},
+    // clock_time_get
+    // fd_read
+    {"wasi_snapshot_preview1", "fd_read",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_read},
+    {"wasi_snapshot_preview1", "fd_write",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_write},
+    {"wasi_snapshot_preview1", "fd_prestat_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_prestat_get},
+    {"wasi_snapshot_preview1", "fd_prestat_dir_name",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32}, fizzy::ValType::i32,
+        wasi_return_enosys},
+    {"wasi_snapshot_preview1", "environ_sizes_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_environ_sizes_get},
+    {"wasi_snapshot_preview1", "environ_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_return_enosys},
+
+
+    {"wasi_unstable", "proc_exit", {fizzy::ValType::i32}, std::nullopt, wasi_proc_exit},
+    // clock_time_get
+    // fd_read
+    {"wasi_unstable", "clock_time_get",
+        {fizzy::ValType::i32, fizzy::ValType::i64, fizzy::ValType::i32}, fizzy::ValType::i32,
+        wasi_return_enosys},
+    {"wasi_unstable", "fd_read",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_read},
+    {"wasi_unstable", "fd_write",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_write},
+    {"wasi_unstable", "fd_prestat_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_fd_prestat_get},
+    {"wasi_unstable", "fd_prestat_dir_name",
+        {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32}, fizzy::ValType::i32,
+        wasi_return_enosys},
+    {"wasi_unstable", "environ_sizes_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_environ_sizes_get},
+    {"wasi_unstable", "environ_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+        fizzy::ValType::i32, wasi_return_enosys},
+};
+
+bool run(int argc, char** argv)
+{
+    uvwasi_options_t options = {
```

Comment:
```suggestion
    const uvwasi_options_t options = {
```

Reply:
`uvwasi_init` is not const: `uvwasi_errno_t uvwasi_init(uvwasi_t* uvwasi, uvwasi_options_t* options);`

Reply:
Add comment then.

Reply:
Created a PR upstream: https://github.com/cjihrig/uvwasi/pull/155